File: playground/go/days/day02/atomic_record.go
package day02

// Day 2 -- API Guarantees: Atomicity.
//
// A record has two fields, A and B. Set() must be all-or-nothing: if
// a + b < 0 the update is invalid and must be REJECTED IN FULL -- neither
// field may change, whether the key already existed or not. Applying one
// field before validating the other is exactly the bug this catches.

type Fields struct {
	A int64
	B int64
}

type RecordStore struct {
	store map[string]Fields
}

func NewRecordStore() *RecordStore {
	return &RecordStore{store: make(map[string]Fields)}
}

// TODO: if a + b >= 0, commit BOTH fields together and return true.
// Otherwise leave the store completely unchanged and return false.
func (r *RecordStore) Set(key string, a, b int64) bool {
	return false
}

// TODO: return the record and true if key has been successfully set, or
// the zero value and false otherwise.
func (r *RecordStore) Get(key string) (Fields, bool) {
	return Fields{}, false
}

File: playground/python/days/day02/atomic_record.py
"""Day 2 -- API Guarantees: Atomicity.

A record has two fields, a and b. A record is valid only when
a + b >= 0.

set(key, a, b) must be all-or-nothing:
  - If a + b >= 0: store both fields together and return True.
  - If a + b < 0: the update is invalid. Reject the WHOLE update -- do
    not change either field -- and return False. This also holds when
    the key already has a record: the old record must stay exactly as
    it was.

get(key) returns the stored record, or None if this key was never
successfully set.
"""

from dataclasses import dataclass
from typing import Optional


@dataclass
class Record:
    a: int
    b: int


class RecordStore:
    def __init__(self) -> None:
        # TODO: choose your own representation.
        pass

    def set(self, key: str, a: int, b: int) -> bool:
        # TODO
        return False

    def get(self, key: str) -> Optional[Record]:
        # TODO
        return None
